src/crafting.h

```diff
@@ -104,6 +104,84 @@ struct recipe {
 
 };
 
+/**
+*   enum used by comp_selection to indicate where a component should be consumed from.
+*/
+enum usage {
+    use_from_map = 1,
+    use_from_player = 2,
+    use_from_both = 1 | 2,
+    use_from_none = 4,
+    cancel = 5 // FIXME: hacky.
+};
+
+/**
+*   Struct that represents a selection of a component for crafting.
+*/
+template<typename CompType = component>
+struct comp_selection {
+    /** Tells us where the selected component should be used from. */
+    usage use_from = use_from_none;
+    CompType comp;
+
+    /** provides a translated name for 'comp', suffixed with it's location e.g '(nearby)'. */
+    std::string nname() const;
+};
+using item_selection = comp_selection<item_comp>;
+using tool_selection = comp_selection<tool_comp>;
+
+/**
+*   Class that describes a crafting job.
+*
+*   The class has functions to execute the crafting job.
+*/
+class craft_command {
+    public:
+        const recipe *rec = nullptr;
```

Comment:
That's usually what I do. I'm still getting used to the to the c++ syntax of public/private blocks instead of access qualifiers for every declaration. It's common practice to put fields at the top of a class in java, so they ended up in the public block in my code.

It's nice to see we're on the same page about that sort of thing though :)